### DISCUSSION TITLE
An ENUM Endpoint API/Metadata  is really necessary

### DISCUSSION THREAD

**Discussion by ManiMozaffar** [fastapi/fastapi#10156] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.post("/classroom/{classroom_id}/student", response_model=StudetnCreationOutput)
async def create_student(
    classroom_id: ClassRoomId,
    token: TokenData = Depends(check_auth),
    database: Type[Database] = Depends(get_database),
    permission_callable: Callable = Depends(get_permission_callable),
    permission_setting: dict[RoutesEnum, dict[UserRoles, PermissionGrants]] = Depends(
        get_permission_setting
    ),
    body: StudetnInput = Body(),
):
    await permission_callable(
        user_rule=token.role,
        user_id=token.user_id,
        route=RoutesEnum.ADD_STUDENT,
        permission_setting=permission_setting,
        classroom_id=classroom_id,
    )
    student = StudentController(database).add_student(
        **body.model_dump(), classroom_id=classroom_id
    )
    return StudetnCreationOutput(student_id=student)
```

### Description

This is my code (very simplified, no db and nothing), checkout ACL folder inside the project...
https://github.com/ManiMozaffar/fast-student/tree/main/fast_acl

I have created an ACL where each Router was defined to have an access level. 
but because that enum is only dependent to that router, I could not perform DI, it was a static parameter that I had to inject to each endpoint!
and then, i had to duplicate that permission_callable in every router...

so basically, it'd be awesome to have some way to access the route's metadata, or whatever I initiated in router, which is only tied to that specific route. in this case, I simply needed a " STATIC READABLE  & RELIABLE NAME", which I call it enum :D

so I think it really make sense to allow us define an ENUM for api, to be optional, then we could catch it later with a type signature. I tried to do that with subclassing ApiRouter, almost succeed, but then I'd lose my backward compatibility on every changes you perform there, so maybe this can be an optional thing inside API?

Note:
- While I can use path, but since they are not hard coded and they can change and they need regex and etc, so I don't think that's a good solution. enum really make sense.
- While I can avoid code duplication with middleware, but again I think it'd be nice to have an awareness inside router to see which route is executed, with a readable code like an Enum. Also, request path may always change in software releases, but enum would remain same.

### Operating System

macOS

### FastAPI Version

0.100.0

### Pydantic Version

2.3

### Python Version

3.11

**Comment by ManiMozaffar** [comment]
I can make a PR myself, and apply my changes. it's almost done, what do you think? worths the changes/efforts?

@tiangolo

**Comment by ManiMozaffar** [comment]
Let's rework this with meta data ?
what do you say?

```python
from fastapi import Metadata
from my_module.config import CustomMetadata

def create_something(metadata: CustomMetadata = Metadata()):
     ...
```

**Comment by tiangolo** [comment, marked as answer]
I'm struggling to understand the use case that you need to solve, maybe I'm not being able to tie the code example with the description. :thinking: 

Could you please create a minimal, self contained, example code that I can copy paste and shows the problem you're having?
